### PR TITLE
Support displaying complex mime types in Output Widget

### LINKED ISSUE #9503: Output messages sent from widgets not displayed in cell output
Hi, 

My system:

``` bash
Version: 1.65.2 (Universal)
Commit: c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1
Date: 2022-03-10T14:33:49.188Z
Electron: 13.5.2
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Darwin x64 21.1.0
```

``` bash
 ➜  ~ python --version
Python 3.10.0
 ➜  ~ jupyter --version
Selected Jupyter core packages...
IPython          : 8.0.1
ipykernel        : 6.9.0
ipywidgets       : 7.7.0
jupyter_client   : 7.1.2
jupyter_core     : 4.9.1
jupyter_server   : not installed
jupyterlab       : not installed
nbclient         : 0.5.13
nbconvert        : 6.4.4
nbformat         : 5.2.0
notebook         : 6.4.10
qtconsole        : 5.2.2
traitlets        : 5.1.1
```

Similar issues: 
- https://github.com/microsoft/vscode-jupyter/issues/3767
- https://github.com/jupyter-widgets/ipywidgets/issues/2512

Code used: 

``` python
from IPython.display import display
from ipywidgets import widgets, interactive
button = widgets.Button(description="Run!")
def on_button_clicked(b):
    print('Clicked')
display(button)
button.on_click(on_button_clicked)
```

In Jupyter Web interface this works perfectly. In VScode it does nothing. 

However, the example give here https://github.com/microsoft/vscode-jupyter/issues/3767#issuecomment-625555606 works just fine. So it seems that they only work with decorators? 

I tested this using:

``` python
from IPython.display import display
from ipywidgets import widgets, interactive, interact
button = widgets.Button(description="Run!")
@interact(b=[0,1])
def on_button_clicked(b):
    if b ==1:
        print('Clicked')
``` 

This seemed to work fine. A bit of a hacky workaround, particularly as decorators don't seem to have a method for buttons...

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #9503

### CODE REVIEW

File: src/notebooks/execution/cellExecutionMessageHandler.ts [hunk 1]
```diff
@@ -30,7 +30,6 @@ import { CellExecutionCreator } from './cellExecutionCreator';
 import { IApplicationShell } from '../../platform/common/application/types';
 import { disposeAllDisposables } from '../../platform/common/helpers';
 import { traceError, traceWarning } from '../../platform/logging';
-import { RefBool } from '../../platform/common/refBool';
```
Comment (DonJayamanne) [inline src/notebooks/execution/cellExecutionMessageHandler.ts:33]: Don't need such a class, all we need is a property (added comments what the property does with reference to Jupyter docs)

File: src/kernels/execution/cellExecutionMessageHandler.ts [hunk 2]
```diff
@@ -376,10 +512,127 @@ export class CellExecutionMessageHandler implements IDisposable {
...
+     * What we do in such a case is append this complex output just after the Output widget.
+     *
+     * In other cases, when appending such output, we need to clear the previous output that we added.
+     *
+     * This method is responsible for:
+     * - Finding the cell that owns the output widget
+     * - Clearing the custom output that was appended after the output widget.
+     * - Appending custom output after the output widget
+     */
+    private updateJupyterOutputWidgetWithOutput(
```
Comment (DonJayamanne) [inline src/kernels/execution/cellExecutionMessageHandler.ts:557]: Todo: Should add more comments

**Review by rchiodo — APPROVED** [review]

**Comment by DonJayamanne** [comment]
>  to grok your algorithm. I just hope we have enough tests.

Yes, I think we have enough. Wil test a little bit more and look for different output widget usages and add more tests.

**Comment by DonJayamanne** [comment]
@rchiodo I'm going to merge this and will submit feature request (discuss with Rob & Matt) for the enhancements.

**Pull request merged**